source/common/network/win32_socket_handle_impl.cc

```diff
@@ -71,10 +79,35 @@ Api::IoCallUint64Result Win32SocketHandleImpl::recvmmsg(RawSliceArrays& slices,
 }
 
 Api::IoCallUint64Result Win32SocketHandleImpl::recv(void* buffer, size_t length, int flags) {
+  if (flags & MSG_PEEK) {
+    // The caller is responsible for calling with the larger size
+    // in cases it needs to do so it can't rely on transparent event activation.
+    // So no in this case we should activate read again unless the read blocked.
+    Api::IoCallUint64Result peek_result = drainToPeekBuffer(length);
```

Comment:
There could be a check for 'length  < peek_buffer_.length()', then skip the drain from the real socket.